src/main/java/com/hivemq/persistence/clientsession/ClientSessionSubscriptionPersistenceImpl.java

```diff
@@ -145,23 +147,17 @@
 
                 final Subscription subscription = new Subscription(sharedTopic, SubscriptionFlags.getDefaultFlags(true, topic.isRetainAsPublished(), topic.isNoLocal()), sharedSubscription.getShareName());
 
-                invalidateSharedSubscriptionCacheAndPoll(client, ImmutableSet.of(subscription));
+                persistFuture = singleWriter.submit(client, (bucketIndex, queueBuckets, queueIndex) -> {
+                    localPersistence.addSubscription(client, topic, timestamp, bucketIndex);
+                    invalidateSharedSubscriptionCacheAndPoll(client, ImmutableSet.of(subscription));
+                    return null;
+                });
             }
 
-            final ListenableFuture<Void> persistFuture = singleWriter.submit(client, (bucketIndex, queueBuckets, queueIndex) -> {
-                localPersistence.addSubscription(client, topic, timestamp, bucketIndex);
-                return null;
-            });
-
             //set future result when local persistence future and topic tree future return;
-            if (sharedSubscription == null) {
-                return Futures.whenAllComplete(persistFuture).call(() -> new SubscriptionResult(topic, subscriberExisted, null),
-                        MoreExecutors.directExecutor());
-            } else {
-                return Futures.whenAllComplete(persistFuture)
-                        .call(() -> new SubscriptionResult(new Topic(sharedSubscription.getTopicFilter(), topic.getQoS()), subscriberExisted, sharedSubscription.getShareName()),
-                                MoreExecutors.directExecutor());
-            }
+            return Futures.whenAllComplete(persistFuture)
+                    .call(() -> new SubscriptionResult(topic, subscriberExisted, sharedSubscription == null ? null : sharedSubscription.getShareName()),
```

Comment:
Is it desired that this now returns "$share/group/topic" instead of "topic" like before ?

Reply:
yes it is, the topic string is not needed anymore, as we need this result only for the retained message sender, which does not apply to shared subs.